front/lib/api/files/upload.ts

```diff
@@ -502,12 +572,165 @@ export async function uploadToCloudStorage(
+      );
+    }
+
+    logger.info(
+      {
+        fileModelId: file.id,
+        workspaceId: auth.workspace()?.sId,
+        infos: infos,
+      },
+      "JIT processing not possible for file."
+    );
+  }
+
+  // Upsert to the conversation datasource & generate a snippet.
+  if (useJit) {
+    const r = await getConversation(auth, file.useCaseMetadata.conversationId);
+
+    if (r.isErr()) {
+      await file.markAsFailed();
+      return new Err({
+        name: "dust_error",
+        code: "internal_server_error",
+        message: `Failed to fetch conversation : ${r.error}`,
+      });
+    }
+
+    const conversation: ConversationType = r.value;
+
+    // Fetch the datasource linked to the conversation...
+    let dataSource = await DataSourceResource.fetchByConversationId(
+      auth,
+      conversation.id
+    );
+
+    if (!dataSource) {
+      // ...or create a new one.
+      const conversationsSpace =
+        await SpaceResource.fetchWorkspaceConversationsSpace(auth);
+
+      // IMPORTANT: never use the conversation sID in the name or description, as conversation sIDs are used as secrets to share the conversation within the workspace users.
+      const name = `Conversation ${uuidv4()}`;
+      const r = await createDataSourceWithoutProvider(auth, {
+        plan: auth.getNonNullablePlan(),
+        owner: auth.getNonNullableWorkspace(),
+        space: conversationsSpace,
+        name: name,
+        description: "Files uploaded to conversation",
+        conversation: conversation,
+      });
+
+      if (r.isErr()) {
+        await file.markAsFailed();
+        return new Err({
+          name: "dust_error",
+          code: "internal_server_error",
+          message: `Failed to create datasource : ${r.error}`,
+        });
+      }
+
+      dataSource = r.value.dataSource;
+    }
+
+    const documentId = file.sId; // Use the file id as the document id to make it easy to track the document back to the file.
+    const sourceUrl = file.getPublicUrl(auth);
+
+    // TODO(JIT) note, upsertDocument do not call runPostUpsertHooks (seems used for document tracker)
+    const upsertDocumentRes = await upsertDocument({
+      name: documentId,
+      source_url: sourceUrl,
+      text: content,
+      parents: [documentId],
+      tags: [`title:${file.fileName}`],
+      light_document_output: true,
+      dataSource,
+      auth,
+    });
+
+    if (upsertDocumentRes.isErr()) {
+      await file.markAsFailed();
+      return new Err({
+        name: "dust_error",
+        code: "internal_server_error",
+        message: "There was an error upserting the document.",
+        data_source_error: upsertDocumentRes.error,
+      });
+    }
+
+    const isTable = await isValidTableFile(file, content);
+
+    if (isTable) {
+      const tableId = file.sId; // Use the file sId as the table id to make it easy to track the table back to the file.
+      const upsertTableRes = await upsertTable({
```

Comment:
Is there any reason not to handle this asynchronously? The file upload calls might become quite slow. Although ideally, you'd want to know the file's state once it returns, I would still separate it from the upload path and treat it as post-processing.

With:
- pre-processing: required to expose to the model
- post-processing: required to handle model action

Reply:
I wanted to do that initially but I was looking to avoid re-reading the file.
In the end, I will do that as it will also solve the edge case of the attachment when creating the convo.